V2 paging

mwbenowitz opened the pull request:
This implements a full set of paging options for the SFR search API. It should allow retrieval for all pages of records in any size of result set (even a search for all records). This builds on the default `from/size` and `search_after` paging options provided by ElasticSearch to construct a full set of options. To support these features several new options have been added as optional parameters to both the `GET` and `POST` endpoints:

- `prev_page_sort` Accepts the `sort` object from a `hit` in the Search API response and returns the previous page of search results. Should always be passed the FIRST `sort` option in a set of results to properly page.
- `next_page_sort` Functions in a similar fashion as the option above, but returns the next page of search results and should always be passed the LAST `sort` option in a set of results.
- `total` The total number of records returned for a search. This is optional but omitting this from a paging requests necessitates retrieving the number from ElasticSearch, slowing the search response.

mwbenowitz committed:
SFR-350 Change swagger documentation from sfr to research-now

Additionally lint minor issues to allow for passing tests

mwbenowitz committed:
SFR-369 Implement paging

To implement a full paging solution for the search API, several
limitations of the ElasticSearch paging structure have to be
worked around. The most significant limitation is that standard
`from/size` paging will only work up through 10,000 results by
default. In addition, paging requests can be expensive as the
position in the result set increases. To address this several
different modes of result page fetching have been implemented:

- For requests up to 10,000 records deep standard paging is
implemented using from/size
- For a request for results within 10,000 records from the end
of the result set. The applied `sort` is reversed, the desired
page retrieved and then has its results reversed again to match
the original sort order
- To retrieve `previous` and `next` pages, the `search_after`
functionality is used. This is a feature of ElasticSearch that allows
for quick retrieval of adjacent results utilizing a `sort` object
that exists on each search result
- To retrieve arbitrary pages that exist in large result sets between
the 10,000 records at either end, a staged query is made. The intial
set, up to the max of 10,000 is retrieved, and from the last record a
series of `search_after` requests are made until the desired page is
in range and can be returned. For performance these "invisible" queries
are made without returning the `_source` object and as a result perform
fairly quickly. For an especially deep paging request (say for 50k-100k
in a result set) this can take up to 5 seconds, but these requests
should be exceedingly rare. If they are not, this will have to be
re-thought, perhaps using the ElasticSearch `scroll` API

To implement these features the search methods have been turned into a
class `Search` that allows for easier construction of the ElasticSearch
query body. The main logic for constructing the query remains, with
additional helpers to manage paging and other features.

mwbenowitz committed:
SFR-369 Lint v2 search and other files

mwbenowitz committed:
SFR-369 Add documentation and additional linting to files

mwbenowitz committed:
SFR-369 Additional linting/spellchecking

mwbenowitz committed:
SFR-369 Add paging options to Swagger docs

nachocs commented:
There is no eslint configuration, so eslint is not running. For example the lack of ';' can be dangerous if you bundle it after.

mwbenowitz committed:
SFR-369 Add eslint config file

mwbenowitz committed:
SFR-369 Refactor Search class

Move the `Search` class into its own file in the `lib` directory
for better separation of concerns and ease of maintainability.

Refactor a few associated files and remove the unused `client.js`
file.

mwbenowitz committed:
SFR-369 Lint all remaining files in repo

mwbenowitz committed:
SFR-369 Updates to fix glitches in newly refactored code

mwbenowitz committed:
SFR-369 Fix loading order for dotenv
